src/stacked_borrows/stack.rs

```diff
@@ -0,0 +1,307 @@
+                    }
+                }),
+            SbTag::Tagged(_) => {
+                #[cfg(feature = "stack-cache")]
+                if let Some(stack_idx) = self.find_granting_cache(access, tag) {
+                    return Some(stack_idx);
+                }
+                // If we didn't find the tag in the cache, fall back to a linear search of the
+                // whole stack, and add the tag to the stack.
+                for (stack_idx, item) in self.borrows.iter().enumerate().rev() {
+                    if tag == item.tag && item.perm.grants(access) {
+                        #[cfg(feature = "stack-cache")]
+                        self.cache.add(stack_idx, tag);
+                        return Some(stack_idx);
+                    }
+                }
+                None
+            }
+        }
+    }
+
+    #[cfg(feature = "stack-cache")]
+    fn find_granting_cache(&mut self, access: AccessKind, tag: SbTag) -> Option<usize> {
+        if self.borrows.len() > CACHE_LEN {
+            // Search the cache for the tag we're looking up
+            if let Some(cache_idx) = self.cache.tags.iter().position(|t| *t == tag) {
+                let stack_idx = self.cache.idx[cache_idx];
+                // If we found the tag, look up its position in the stack to see if it grants
+                // the required permission
+                if self.borrows[stack_idx].perm.grants(access) {
+                    // If it does, and it's already in the most-recently-used position, move it
+                    // there.
+                    if cache_idx != 0 {
+                        self.cache.add(stack_idx, tag);
+                    }
+                    return Some(stack_idx);
+                } else {
+                    // Tag is in the cache, but it doesn't grant the required permission
+                    return None;
+                }
+            }
+        }
+        None
+    }
+
+    pub fn insert(&mut self, new_idx: usize, new: Item) {
+        self.borrows.insert(new_idx, new);
+
+        #[cfg(feature = "stack-cache")]
+        self.insert_cache(new_idx, new);
+    }
+
+    #[cfg(feature = "stack-cache")]
+    fn insert_cache(&mut self, new_idx: usize, new: Item) {
+        // Adjust the possibly-unique range if an insert occurs before or within it
+        if self.unique_range.start >= new_idx {
+            self.unique_range.start += 1;
+        }
+        if self.unique_range.end >= new_idx {
+            self.unique_range.end += 1;
+        }
+        if new.perm == Permission::Unique {
+            // Make sure the possibly-unique range contains the new borrow
+            self.unique_range.start = self.unique_range.start.min(new_idx);
+            self.unique_range.end = self.unique_range.end.max(new_idx + 1);
+        }
+
+        // The above insert changes the meaning of every index in the cache >= new_idx, so now
+        // we need to find every one of those indexes and increment it.
```

Comment:
This makes me wonder if we should do a bit more work to do *less* work when the borrow stack is small -- smaller than the cache, even. It seems to me like this cache will massively slow down operations on stacks that are about as large as the cache, due to all the extra bookkeeping.

OTOH your benchmarks show a clear speedup, so I guess the lesson is that the cost is so dominated by the large stacks that making the small stacks slower is fine? Or that this is all O(1) and complexity theory is actually right? Or something like that?

Reply:
First off, I agree that this loop is a performance hazard. I think I mentioned that at your request in the documentation of `CACHE_LEN`.

I am pretty sure we get away with this behavior because the cache is a crazy efficient data structure, especially by comparison to the rest of Miri. It's just loops over a small(ish) array of constant length and modern processors love code like that. It _is_ O(1), but I hesitate to use that as a justification for anything because not all programs push Stacked Borrows into complexity theory territory. For example, if a program somehow had a totally pathological borrowing pattern it could end up with mostly cache misses and then we would really see these O(1) cache operations surface. But to be clear, I have yet to find such a program. I checked in the most pathological programs I could find as other benchmarks, and _at worst_ these cache operations amount to 2.2% of Miri's runtime. Which is minuscule compare to what runtime the cache is saving.

This is so messy from a theoretical perspective, and we should really have better benchmarking (something like a fork of rustc-perf would be cool), but also this is such a dramatic win on so many programs that I would prefer to land this optimization, then do a part 2 with more unambiguous but smaller wins.

Reply:
That sounds totally fair, I just wanted to be sure. :) 

If you could summarize this in one sentence for a comment in the code, that would be great. Basically just acknowledging that yes this might seem like a waste for tiny stacks but in practice it is a huge win since (a) "modern processors love code like that" and (b) runtime is dominated by large stacks.